Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -4860,68 +4860,89 @@ - (void)reloadViewItems
     }
 
-    // Update the "shouldShowDate" property of the view items.
+    // Update the properties of the view items.
     //
-    // First iterate in reverse order.
-    OWSInteractionType lastInteractionType = OWSInteractionType_Unknown;
-    MessageReceiptStatus lastReceiptStatus = MessageReceiptStatusUploading;
-    NSString *_Nullable lastIncomingSenderId = nil;
-    for (ConversationViewItem *viewItem in viewItems.reverseObjectEnumerator) {
+    // NOTE: This logic uses shouldShowDate which is set in the previous pass.
+    for (NSUInteger i = 0; i < viewItems.count; i++) {
+        ConversationViewItem *viewItem = viewItems[i];
+        ConversationViewItem *_Nullable previousViewItem = (i > 0 ? viewItems[i - 1] : nil);
+        ConversationViewItem *_Nullable nextViewItem = (i + 1 < viewItems.count ? viewItems[i + 1] : nil);
         BOOL shouldShowSenderAvatar = NO;
         BOOL shouldHideFooter = NO;
+        NSString *_Nullable senderName = nil;
 
         OWSInteractionType interactionType = viewItem.interaction.interactionType;
+        NSString *timestampText = [DateUtil formatTimestampShort:viewItem.interaction.timestamp];
 
         if (interactionType == OWSInteractionType_OutgoingMessage) {
             TSOutgoingMessage *outgoingMessage = (TSOutgoingMessage *)viewItem.interaction;
             MessageReceiptStatus receiptStatus =
                 [MessageRecipientStatusUtils recipientStatusWithOutgoingMessage:outgoingMessage
                                                                   referenceView:self.view];
 
-            // Always show "failed to send" status.
-            shouldHideFooter = (interactionType == lastInteractionType && receiptStatus == lastReceiptStatus
-                && outgoingMessage.messageState != TSOutgoingMessageStateFailed);
-
-            lastReceiptStatus = receiptStatus;
+            if (nextViewItem && nextViewItem.interaction.interactionType == interactionType) {
+                TSOutgoingMessage *nextOutgoingMessage = (TSOutgoingMessage *)nextViewItem.interaction;
+                MessageReceiptStatus nextReceiptStatus =
+                    [MessageRecipientStatusUtils recipientStatusWithOutgoingMessage:nextOutgoingMessage
+                                                                      referenceView:self.view];
+                NSString *nextTimestampText = [DateUtil formatTimestampShort:nextViewItem.interaction.timestamp];
+
+                // We can skip the "outgoing message status" footer if the next message
+                // has the same footer and no "date break" separates us...
+                // ...but always show "failed to send" status.
+                shouldHideFooter = ([timestampText isEqualToString:nextTimestampText]
+                    && receiptStatus == nextReceiptStatus
+                    && outgoingMessage.messageState != TSOutgoingMessageStateFailed && !nextViewItem.shouldShowDate);
```

Comment:
We don't want to collapse sender names/sender avatars/footers across "date breaks".